hal/src/thumbv7em/clock/v2.rs

```diff
@@ -1,98 +1,540 @@
+//! - Some of the [`ApbClks`]
+//! - The 48 MHz [`Dfll`], running in [`OpenLoop`] mode
+//! - [`Gclk0`], sourced by the `Dfll`, which feeds the processor's master clock
+//! - The [`OscUlpBase`] clock, which is always running and cannot be disabled.
+//!   It can feed the derived [`OscUlp1k`] and [`OscUlp32k`] clocks.
+//!
+//! Note that bus clocks are not wrapped in [`Enabled`], because they can never be
+//! used to drive downstream clocks. The other three clocks, on the other hand,
+//! are all wrapped in [`Enabled`].
+//!
+//! The [`Dfll`] is configured as `Enabled<Dfll<OpenLoop>, U1>`, which
+//! represents fact that [`Gclk0`] is dependent on it. [`Gclk0`] is configured
+//! as `Enabled<Gclk0<DfllId>, U1>`, which indicates that it is derived from the
+//! [`Dfll`] (note the use of [`DfllId`] as an `Id` type) and has one dependent
+//! clock of its own. The dependent clock is, of course, the processor master
+//! clock.
+//!
+//! [`Gclk0`] can be configured to use a different base clock, which would allow
+//! you to reduce the [`Dfll`] [`Enabled`] [`Counter`] from [`U1`] to [`U0`] and
+//! disable it. However, the same is not true for [`Gclk0`]. Because the master
+//! clock has no representation in this module, the [`Counter`] for
+//! [`EnabledGclk0`] can *never* be reduced to [`U0`], which neatly models that
+//! the master clock can never be disabled.
+//!
+//! Finally, the [`OscUlpBase`] clock is modelled as `Enabled<OscUlpBase, U0>`,
+//! because it has no dependent clocks at power-on reset. However, it can never
+//! be disabled, so we do not provide any `.disable()` method.
+//!
+//! The [`Tokens`] struct contains all of the available `Token`s, which
+//! represent clocks that are disabled at power-on reset. Each `Token` can be
+//! exchanged for a corresponding clock object.
+//!
+//! ## Example clock tree
+//!
+//! Finally, we will walk through the creation of a simple clock tree to
+//! illustrate some of the remaining concepts inherent to this module.
+//!
+//! Starting from the previous snippet, we have the [`Buses`], [`Clocks`] and
+//! [`Tokens`] to work with, and our clock tree at power-on reset looks like
+//! this.
+//!
+//! ```text
+//! DFLL (48 MHz)
+//! └── GCLK0 (48 MHz)
+//!     └── Master clock (48 MHz)
+//!
+//! OSCULP base clock
+//! ```
+//!
+//! Our goal will be a clock tree that looks like this:
+//!
+//! ```text
+//! XOSC0 (8 MHz)
+//! └── DPLL0 (100 MHz)
+//!     └── GCLK0 (100 MHz)
+//!         ├── Master clock (100 MHz)
+//!         ├── SERCOM0 peripheral clock
+//!         └── Ouput to GPIO pin
+//!
+//! OSCULP base clock
+//! ```
+//!
+//! We will use an external crystal oscillator running at 8 MHz to feed a DPLL,
+//! which will increase the clock frequency to 100 MHz. Then, we will
+//! reconfigure GCLK0 to use the 100 MHz DPLL clock instead of the 48 MHz DFLL
+//! clock.
+//!
+//! First, let's import some of the necessary types. We will see what each type
+//! represents in turn.
+//!
+//! ```
```

Comment:
I was going to do that, but I was worried the docs would get huge within the code, all for a bunch of repeat copies of the same code commented out. But you make a good argument about stale documentation and doc tests.

Reply:
Not sure what can be done about it. It'd be nice to have like a "common hidden" code shared between code blocks. Like, let's say, imports and clock struct initialization and then you can have short example code blocks reusing it. Maybe I missed something like that in Cargo book.

Reply:
I've made all the examples in the main documentation `no_run` instead of `ignore`. I'll leave this open until I update the `gclk`, `dpll` and `pclk` modules.